### PR TITLE
feat: add `Series|Expr.cum_prod` method, add `reverse` kw in `cum_sum` method

### PR DESCRIPTION
**Pull request by FBruzzesi**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues

- Related issue #1371 

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added
- [x] Documented the changes

## If you have comments or can explain your changes, please do so below

This is 3 folded (I could not resist)

- Introduce `cum_prod`
- Adds `reverse` in `cum_sum`
- Adds cumulative methods for dask if `reverse=False`, raise otherwise

### CODE REVIEW

File: narwhals/_arrow/series.py [hunk 1]
```diff
@@ -815,12 +819,7 @@ def mode(self: Self) -> ArrowSeries:
         )[self.name]
 
     def cum_count(self: Self, *, reverse: bool) -> Self:
-        not_na_series = (~self.is_null()).cast(self._dtypes.UInt32())
-        return (
-            not_na_series.cum_sum()
-            if not reverse
-            else len(self) - not_na_series.cum_sum() + not_na_series - 1
-        )
```
Comment (FBruzzesi) [inline narwhals/_arrow/series.py:823]: Ok maybe this could be faster as it does not reverse the array twice

File: narwhals/expr.py [hunk 2]
```diff
@@ -2697,6 +2700,9 @@ def cum_count(self: Self, *, reverse: bool = False) -> Self:
         Arguments:
             reverse: reverse the operation
 
+        Returns:
+            A new expression.
+
```
Comment (FBruzzesi) [inline narwhals/expr.py:2705]: Guess who forgot about these _returns_

Comment (MarcoGorelli) [reply narwhals/expr.py:2705]: 😆 https://github.com/narwhals-dev/narwhals/pull/1387 will enforce them

**Review by MarcoGorelli — APPROVED** [review]
thanks @FBruzzesi !

**Pull request merged**